src/bpfbytecode.cpp

```diff
@@ -25,4 +26,89 @@ const std::vector<uint8_t> &BpfBytecode::getSection(
   return sections_.at(name);
 }
 
+// This is taken from libbpf (btf.c) and we need it to manually iterate BTF
+// entries so that we can fix them up in place.
+// Should go away once we let libbpf do the BTF fixup.
+static int btf_type_size(const struct btf_type *t)
+{
+  const int base_size = sizeof(struct btf_type);
+  __u16 vlen = btf_vlen(t);
+
+  switch (btf_kind(t))
+  {
+    case BTF_KIND_FWD:
+    case BTF_KIND_CONST:
+    case BTF_KIND_VOLATILE:
+    case BTF_KIND_RESTRICT:
+    case BTF_KIND_PTR:
+    case BTF_KIND_TYPEDEF:
+    case BTF_KIND_FUNC:
+    case BTF_KIND_FLOAT:
+    case BTF_KIND_TYPE_TAG:
+      return base_size;
+    case BTF_KIND_INT:
+      return base_size + sizeof(__u32);
+    case BTF_KIND_ENUM:
+      return base_size + vlen * sizeof(struct btf_enum);
+    case BTF_KIND_ENUM64:
+      return base_size + vlen * sizeof(struct btf_enum64);
+    case BTF_KIND_ARRAY:
+      return base_size + sizeof(struct btf_array);
+    case BTF_KIND_STRUCT:
+    case BTF_KIND_UNION:
+      return base_size + vlen * sizeof(struct btf_member);
+    case BTF_KIND_FUNC_PROTO:
+      return base_size + vlen * sizeof(struct btf_param);
+    case BTF_KIND_VAR:
+      return base_size + sizeof(struct btf_var);
+    case BTF_KIND_DATASEC:
+      return base_size + vlen * sizeof(struct btf_var_secinfo);
+    case BTF_KIND_DECL_TAG:
+      return base_size + sizeof(struct btf_decl_tag);
+    default:
+      return -EINVAL;
+  }
+}
+
+// There are cases when BTF generated by LLVM needs to be patched. This is
+// normally done by libbpf but only when loading via bpf_object is used. We're
+// currently using direct loading using bpf_btf_load and bpf_prog_load so we
+// need to mimic libbpf's behavior and do the patching manually.
+// This should go away once we move to bpf_object-based loading.
+//
+// Transformations done:
+// - If running on a kernel not supporting BTF func entries with BTF_FUNC_GLOBAL
+//   linkage, change the linkage type to 0. We cannot do this directly in
+//   codegen b/c LLVM would optimize our functions away.
+void BpfBytecode::fixupBTF(BPFfeature &feature)
+{
+  if (!hasSection(".BTF"))
+    return;
```

Comment:
Should there be an error log or throw here?

Reply:
I don't think so. If there's no BTF, there's nothing to fixup. While this wouldn't happen now, we may want to have an option not to generate BTF in future. If BTF is not available while it should be, we'll catch that in other places.

Reply:
Ah ok, I had thought this was just a validation check.